From 0b16fb20273ba4c0aa744aaa5df808bcfb01d84b Mon Sep 17 00:00:00 2001
From: Prasant Singh <prasant.singh@ittiam.com>
Date: Mon, 11 Dec 2023 17:56:49 +0530
Subject: [PATCH] Fixes for issue 2131

The sad8xhx4d_neon function assumed a minimum block height of 2,
leading to a segmentation fault when encountering blocks with a height
of 1. This function will now accept all heights.
---
 Source/Lib/Encoder/ASM_NEON/EbComputeSAD_neon.c | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/Source/Lib/Encoder/ASM_NEON/EbComputeSAD_neon.c b/Source/Lib/Encoder/ASM_NEON/EbComputeSAD_neon.c
index 7f2d91884..061dd382d 100644
--- a/Source/Lib/Encoder/ASM_NEON/EbComputeSAD_neon.c
+++ b/Source/Lib/Encoder/ASM_NEON/EbComputeSAD_neon.c
@@ -514,15 +514,13 @@ static INLINE uint32x4_t sad8xhx4d_neon(const uint8_t *src, uint32_t src_stride,
     uint8x8_t  s;
     uint32_t   i, ref_offset;
 
-    s      = vld1_u8(src);
-    sum[0] = vabdl_u8(s, vld1_u8(ref + 0));
-    sum[1] = vabdl_u8(s, vld1_u8(ref + 1));
-    sum[2] = vabdl_u8(s, vld1_u8(ref + 2));
-    sum[3] = vabdl_u8(s, vld1_u8(ref + 3));
-
-    src += src_stride;
-    ref_offset = ref_stride;
-    i          = h - 1;
+    sum[0] = vdupq_n_u16(0);
+    sum[1] = vdupq_n_u16(0);
+    sum[2] = vdupq_n_u16(0);
+    sum[3] = vdupq_n_u16(0);
+
+    ref_offset = 0;
+    i          = h;
     do {
         s      = vld1_u8(src);
         sum[0] = vabal_u8(sum[0], s, vld1_u8(ref + 0 + ref_offset));
-- 
GitLab

